### PR TITLE
Fixes #16219, `hasArgOfName` ignoring argument sets.

### LINKED ISSUE #16219: `hasArgOfName` misses the tail of a list of parameter names with the same type
The arguments `j` and `k` are not found:

### Example
```nim
import macros

macro m(u:untyped):untyped =
  echo:
    [ hasArgOfName((params u),"i"),
      hasArgOfName((params u),"j"),
      hasArgOfName((params u),"k")
    ]

proc p(i,j,k:int) {.m.} = discard
```

### Current Output
```
[true, false, false]
```

### Expected Output
```
[true, true, true]
```

### Possible Solution

`echo u.params.len` outputs `2`, the first is the empty node for the function output type so the length of input parameters is wrongly detected as `1`. Since `hasArgOfName` uses it to search the param list it only finds `i`.

```
$ nim -v
Nim Compiler Version 1.5.1 [Linux: amd64]
Compiled at 2020-12-01
Copyright (c) 2006-2020 by Andreas Rumpf

git hash: 1d1b3f79fdccfdd359781dbde3c5aea3b67bf05b
active boot switches: -d:release
```

### PR DESCRIPTION
**Pull request by deech**

### CODE REVIEW

File: tests/macros/thasargofname.nim [hunk 1]
```diff
@@ -0,0 +1,20 @@
+discard """
+  nimout: '''
+  [true, true, true, true, true, true, true]
+  '''
+"""
+
```
Comment (timotheecour) [inline tests/macros/thasargofname.nim]: move this to `tests/stdlib/tmacros.nim`, which is the convention followed by almost all stdlib modules

File: tests/macros/thasargofname.nim [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+import macros
+
+macro m(u:untyped):untyped =
+  echo:
```
Comment (timotheecour) [inline tests/macros/thasargofname.nim]:
```nim
doAssert hasArgOfName((params u), "s")
doAssert hasArgOfName((params u), "i")
```
also need negative tests:
```nim
doAssert not hasArgOfName((params u), "nonexistant")
```

File: tests/macros/thasargofname.nim [hunk 3]
```diff
@@ -0,0 +1,20 @@
+discard """
```
Comment (timotheecour) [inline tests/macros/thasargofname.nim]: not needed if you follow https://github.com/nim-lang/Nim/pull/16233/files#r534416096

File: lib/core/macros.nim [hunk 4]
```diff
@@ -1435,9 +1435,35 @@ proc hasArgOfName*(params: NimNode; name: string): bool {.compileTime.}=
   ## Search ``nnkFormalParams`` for an argument.
   expectKind(params, nnkFormalParams)
   for i in 1 ..< params.len:
-    template node: untyped = params[i]
-    if name.eqIdent( $ node[0]):
-      return true
+    # In, for example, proc p(s:string;i,j,k:int) ..., the i,j and k arguments
```
Comment (timotheecour) [inline lib/core/macros.nim]: I'd remove this whole comment, nothing surprising is going on here (and other more complicated procs don't do that either)

File: lib/core/macros.nim [hunk 5]
```diff
@@ -1435,9 +1435,35 @@ proc hasArgOfName*(params: NimNode; name: string): bool {.compileTime.}=
...
+    #     newIdentNode("k"),
+    #     newIdentNode("int"),
+    #     newEmptyNode()
+    #   )
+    # )
+    # ... so to find an argument name in the i,j,k argument set we have to loop
+    # over the innards of the second nnkIdentDefs. The only way to identify an
+    # argument set is if the nnkIdentDefs node has more than 3 children, two
+    # nodes for the default value and type and the rest are argument names.
+    if params[i].len > 3:
```
Comment (timotheecour) [inline lib/core/macros.nim]: no need to distinguish `> 3` as a special case; this works:
```nim
    for j in 0..params[i].len-3:
      if name.eqIdent($params[i][j]):
        return true
```

**Comment by deech** [comment]
@timotheecour Fixed according to comments.

File: tests/stdlib/tmacros.nim [hunk 6]
```diff
@@ -0,0 +1,13 @@
+import macros
+
+macro m(u:untyped):untyped =
```
Comment (timotheecour) [inline tests/stdlib/tmacros.nim]: space: `macro m(u:untyped): untyped =`

Comment (timotheecour) [inline tests/stdlib/tmacros.nim]: move under a block as in other tests in tests/stdlib/txxx.nim:

```nim
block: # hasArgOfName
  macro m(u:untyped): untyped =
    ...
  proc p(s:string; i,j,k:int; b:bool; xs,ys:seq[int] = @[]) {.m.} = discard
```

makes it clear what proc is being tested, helps when searching where a proc is tested (vs used), and avoids scope pollution when more tests are added to this file

File: tests/stdlib/tmacros.nim [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+  doAssert hasArgOfName(params u,"s")
```
Comment (timotheecour) [inline tests/stdlib/tmacros.nim]:
```nim
for name in ["i", "j", ...]:
  doAssert hasArgOfName(params u, name)
doAssert not hasArgOfName(params u,"nonexistent")
```

**Comment by timotheecour** [comment]
@deech almost there, added a few more comments, then I'll LGTM

**Review by timotheecour — APPROVED** [review]

File: tests/stdlib/tmacros.nim [hunk 8]
```diff
@@ -0,0 +1,9 @@
+import macros
+
+block: # hasArgOfName
+  macro m(u:untyped): untyped =
```
Comment (timotheecour) [inline tests/stdlib/tmacros.nim]: also spacing here: `m(u: untyped): untyped =` (can be fixed in subsequent cleanup, or here)

File: tests/stdlib/tmacros.nim [hunk 9]
```diff
@@ -0,0 +1,9 @@
...
+    for name in ["s","i","j","k","b","xs","ys"]:
+      doAssert hasArgOfName(params u,name)
+    doAssert not hasArgOfName(params u,"nonexistent")
+
+  proc p(s:string; i,j,k:int; b:bool; xs,ys:seq[int] = @[]) {.m.} = discard
```
Comment (timotheecour) [inline tests/stdlib/tmacros.nim]: ditto

Comment (deech) [reply tests/stdlib/tmacros.nim]: I'll fix it here. Is there a formatting tool that automates this?

File: tests/macros/thasargofname.nim (hunk 1 above)
Comment (timotheecour) [reply tests/macros/thasargofname.nim]: @deech  please click "resolve conversation" once you've resolved it :)

File: tests/stdlib/tmacros.nim (hunk 9 above)
Comment (timotheecour) [reply tests/stdlib/tmacros.nim]: `nimpretty` but you still need to double check results (and should only be applied on code you're changing, refs https://github.com/nim-lang/Nim/issues/8458#issuecomment-457055195

Comment (Clyybber) [reply tests/stdlib/tmacros.nim]: It really doesn't matter for tests.

File: lib/core/macros.nim [hunk 10]
```diff
@@ -1435,9 +1435,9 @@ proc hasArgOfName*(params: NimNode; name: string): bool {.compileTime.}=
   ## Search ``nnkFormalParams`` for an argument.
   expectKind(params, nnkFormalParams)
   for i in 1 ..< params.len:
-    template node: untyped = params[i]
-    if name.eqIdent( $ node[0]):
-      return true
+    for j in 0..params[i].len-3:
```
Comment (Clyybber) [inline lib/core/macros.nim:1438]:
```suggestion
    for j in 0..<params[i].len-2:
```

File: lib/core/macros.nim [hunk 11]
```diff
@@ -1435,9 +1435,9 @@ proc hasArgOfName*(params: NimNode; name: string): bool {.compileTime.}=
   ## Search ``nnkFormalParams`` for an argument.
   expectKind(params, nnkFormalParams)
   for i in 1 ..< params.len:
-    template node: untyped = params[i]
-    if name.eqIdent( $ node[0]):
-      return true
+    for j in 0..<params[i].len-2:
```
Comment (ringabout) [inline lib/core/macros.nim:1438]: need space

Comment (Clyybber) [reply lib/core/macros.nim:1438]: Why and where?

Comment (ringabout) [reply lib/core/macros.nim:1438]: The code above uses space, maybe keep consistent:
```nim
for i in 1 ..< params.len:
    for j in 0..<params[i].len-2:
```

Comment (Clyybber) [reply lib/core/macros.nim:1438]: Its more common without spaces and looks better so I'm gonna change the other line too.

File: lib/core/macros.nim [hunk 12]
```diff
@@ -1435,9 +1435,9 @@ proc hasArgOfName*(params: NimNode; name: string): bool {.compileTime.}=
   ## Search ``nnkFormalParams`` for an argument.
   expectKind(params, nnkFormalParams)
   for i in 1 ..< params.len:
```
Comment (Clyybber) [inline lib/core/macros.nim:1437]:
```suggestion
  for i in 1..<params.len:
```

**Pull request merged**